### PR TITLE
ISSUE-410 (v0.3.1) AvroFieldsGenerator produces Stackoverflow error when parsing a nested record schema.

### PR DESCRIPTION
**Pull request by Wesley-Lawrence**
Added test resources for testing nested schemas.
Added tests to make sure AvroFieldsGenerator and AvroSchemaProvider can parse nested schemas.
Added code to AvroFieldsGenerator to handle parsing nested schemas.

I'd like to fix this against v0.3.1, since we're looking at using the Hortonworks Schema Registry with NiFi, and as of this PR submission, NiFi 1.5 is compatible with v0.3.0, and it seems there was API changes in v0.4+

### CODE REVIEW

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/avro/AvroFieldsGenerator.java [hunk 1]
```diff
@@ -67,36 +70,48 @@ private void parseField(Schema.Field field, List<SchemaFieldInfo> schemaFieldInf
...
-                // store fields of a record.
-                List<Schema.Field> fields = schema.getFields();
-                for (Schema.Field recordField : fields) {
-                    parseField(recordField, schemaFieldInfos);
+
+                String completeName = schema.getFullName();
+
+                if (visitedRecords.contains(completeName)) {
+                    // It seems like the goal of this class is to only return a list of all primitive fields.
+                    //   If we encounter a record that was already parsed, then just ignore it...
```
Comment (raju-saravanan) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/avro/AvroFieldsGenerator.java]: Could you rephrase the comments to "Since we are only interested in primitive data types, if we encounter a record that was already parsed it can be ignored"?. Beginning the comment with "it seems like" sounds speculative.

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @BeforeClass
+    public static void beforeClass() throws IOException {
+        Schema.Parser schemaParser = new Schema.Parser();
+        simpleNestedSchema = schemaParser.parse(AvroNestedCheckerTest.class.getResourceAsStream("/avro/nested/nested-simple.avsc"));
+        complexNestedSchema = schemaParser.parse(AvroNestedCheckerTest.class.getResourceAsStream("/avro/nested/nested-complex.avsc"));
+    }
+
+    @Test
+    public void testAvroFieldsGenerator_Simple() throws IOException {
+        // As long as no exceptions are thrown, this passes.
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: nit: Could you rename the method to "testSimpleAvroFieldGenerator"?

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        new AvroFieldsGenerator().generateFields(simpleNestedSchema);
+    }
+
+    @Test
+    public void testAvroFieldsGenerator_Complex() throws IOException {
+        // As long as no exceptions are thrown, this passes.
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: nit: Could you rename the method to "testComplexAvroFieldsGenerator"?

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        new AvroFieldsGenerator().generateFields(complexNestedSchema);
+    }
+
+    @Test
+    public void testAvroSchemaProvider_Simple() throws IOException {
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: nit: Could you rename the method to "testSimpleAvroSchemaProvider"?

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        // As long as no exceptions are thrown, this passes.
+        new AvroSchemaProvider().normalize(simpleNestedSchema);
+    }
+
+    @Test
+    public void testAvroSchemaProvider_Complex() throws IOException {
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: nit: Could you rename the method to "testComplexAvroSchemaProvider"?

File: schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          {
+            "name": "parent",
+            "type": [
+              "null",
+              "Record_A"
+            ]
+          }
+        ]
+      }
+    }
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc]: Could you add other data structure like arrays and map which has self-referential types?

File: schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc [hunk 7]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc]: Could you move this file to /resources/avro/composites/ ?

File: schema-registry/common/src/test/resources/avro/nested/nested-simple.avsc [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/resources/avro/nested/nested-simple.avsc]: Could you move this file to /resources/avro/composites/ ?

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package com.hortonworks.registries.schemaregistry.avro;
+
+import org.apache.avro.Schema;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.IOException;
+
+public class AvroNestedCheckerTest {
```
Comment (raju-saravanan) [inline schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: Could you add all the tests in this call to "AvroCompositeSchemasTest"? AvroCompositeSchemasTest class is used for testing complex schemas, the tests in this class would be a good fit there.

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/avro/AvroFieldsGenerator.java (hunk 1 above)
Comment (Wesley-Lawrence) [reply schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/avro/AvroFieldsGenerator.java]: Good point, will change 🙂

File: schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc (hunk 6 above)
Comment (Wesley-Lawrence) [reply schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc]: Good idea, will do 🙂

File: schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc (hunk 7 above)
Comment (Wesley-Lawrence) [reply schema-registry/common/src/test/resources/avro/nested/nested-complex.avsc]: It seemed like that folder was for testing schemas that leverage Hortonworks "includeSchemas" feature, so I thought keeping schemas related to "nesting" somewhere else would keep things more organized.

Let me know what you think; I can change it if you prefer composites and nested kept together.

File: schema-registry/common/src/test/resources/avro/nested/nested-simple.avsc (hunk 8 above)
Comment (Wesley-Lawrence) [reply schema-registry/common/src/test/resources/avro/nested/nested-simple.avsc]: Ditto above.

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java (hunk 9 above)
Comment (Wesley-Lawrence) [reply schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: My only concern here, is that the composite aspect of schemas isn't Avro specific AFAIK; it's something Hortonworks added. Where the ability to nest records inside of them selves is standard Avro.

It seems to me that they're separate problems (even if they sometimes leverage the same code path), and should get tested separately.

Let me know what you think; I can change it so they're grouped together if that's preferred.

**Comment by Wesley-Lawrence** [comment]
@raju-saravanan Thanks for the review! I made most of the changes.

Let me know how you feel about keeping nested and composite separate. While I feel that they should be separate, I can put them together if you feel strongly about it 🙂

File: schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java (hunk 9 above)
Comment (raju-saravanan) [reply schema-registry/common/src/test/java/com/hortonworks/registries/schemaregistry/avro/AvroNestedCheckerTest.java]: @Wesley-Lawrence: Yeah that makes sense. I agree we should keep them separate.

**Comment by raju-saravanan** [comment]
@Wesley-Lawrence: +1 LGTM, thanks a lot for the patch.
